Update flow.to_global/flow.to_local to Support Multiple Input Types

Alive1024 opened the pull request:
根据 https://github.com/Oneflow-Inc/OneTeam/issues/1031 中的讨论为 `flow.to_global/flow.to_local` 增加多种输入类型的输入支持，尤其关注 state dict

测试：
- [x] None
- [x] Tensor
- [x] list
- [x] tuple
- [x] nn.Module 的 state dict
- [x] nn.Graph 的 state dict
- [x] 几种类型的任意组合


**文档预览**：
1. `oneflow.to_global`
![image](https://user-images.githubusercontent.com/46194306/165489391-d10fe4f3-cf22-43d4-9cfb-7b500dfb7203.png)
![image](https://user-images.githubusercontent.com/46194306/165489502-88fe16dd-9ae9-415e-964e-9757bc7a6a9f.png)

2. `oneflow.to_local`
![image](https://user-images.githubusercontent.com/46194306/165489543-90f32c6a-a9fa-47fd-90e9-e818d767b7ae.png)
![image](https://user-images.githubusercontent.com/46194306/165489730-416954ab-d8d2-49e1-9d33-799ea9fd74e1.png)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Oneflow-Inc/oneflow?pullRequest=8091) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Oneflow-Inc/oneflow?pullRequest=8091) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Oneflow-Inc/oneflow?pullRequest=8091) it.</sub>

Alive1024 commented:
在原本的接口中，输入参数名为 `input`，这与 Python 的 built-in 函数 `input` 同名，是否有必要需要改动下？

Alive1024 commented:
@strint @daquexian 两位前辈有空了可以先 review 下目前的代码吗？第一次向 OneFlow 提交 PR，可能有诸多纰漏，请两位前辈多多指教~

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -75,16 +77,56 @@ def global_to_global_op(
-        return local_to_global_op(input=input, placement=placement, sbp=sbp, **kwargs)
+    r"""Converts the input to global if it is local, otherwise performs designated placement and/or sbp conversion.
+    
+    It performs the same conversion as :func:`oneflow.Tensor.to_global` to the tensor or every tensor in the state dict.
```
in the state dict  -> input

Alive1024 replied:
Corrected

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -75,16 +77,56 @@ def global_to_global_op(
+    input_tree = IONode(value=input)
+
+    def leaf_fn(node):
+        if isinstance(node._value, Tensor):
```
Define a `to_global_tensor` function to do this.

if is Tensor, call `to_global_tensor`, else build a `IONode`

Alive1024 replied:
Corrected

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -75,16 +77,56 @@ def global_to_global_op(
-        if "grad_sbp" in kwargs:
-            del kwargs["grad_sbp"]
-        return local_to_global_op(input=input, placement=placement, sbp=sbp, **kwargs)
+    r"""Converts the input to global if it is local, otherwise performs designated placement and/or sbp conversion.
```
Converts the input tensor or input tensor in List/Tuple/Dict to a global tensor.

strint replied:
Refer to flollowing apis to refine the API Doc:
- https://oneflow.readthedocs.io/en/master/tensor.html?highlight=to_global#oneflow.Tensor.to_global
- https://oneflow.readthedocs.io/en/master/tensor.html?highlight=to_global#oneflow.Tensor.to_local

API doc should has an example.

Alive1024 replied:
Corrected, example added

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -75,16 +77,56 @@ def global_to_global_op(
+    input_tree = IONode(value=input)
+
+    def leaf_fn(node):
+        if isinstance(node._value, Tensor):
```
Refer to  `to_global_tensor`, refactor this with `to_local_tensor`.

Alive1024 replied:
Corrected

strint commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -0,0 +1,127 @@
+    local_graph_model = GraphModel(model)
+    global_graph_model = GraphModel(model.to_global(placement=placement, sbp=sbp))
+
+    def _all_global(test_case, input):
```
Add check on placement and sbp

Alive1024 replied:
Added

strint commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -75,16 +77,135 @@ def global_to_global_op(
+        True
+        True
+    """
+    def _to_global_tensor(input_tensor, placement=None, sbp=None, **kwargs):
```
move outside this function to avoid def at every call

Alive1024 replied:
Corrected

clackhan commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+        input_tree = IONode(value=input)
+
+        def leaf_fn(node):
+            if isinstance(node._value, Tensor):
+                return _to_local_tensor(node._value)
+            else:
+                warnings.warn("Non-Tensor type: {} encountered, it will not be converted.".format(type(node._value)))
+                return node._value
```
这里的意思感觉不太对，看起来像是如果本rank输入是None，输出也是None，正常情况应该是所有rank输出都是一样的global tensor。类似下面写法：
```python
# world_size = 2
import oneflow as flow
if flow.env.get_rank() == 0:
    x = flow.tensor([1.0, 2.0, 3.0, 4.0])
else:
    x =None

y = flow.to_global(x, placement=flow.placement("cpu", [1]), sbp=flow.sbp.broadcast)
y.placement
oneflow.placement(type="cpu", ranks=[1]) # rank0 和 rank1 输出相同
y.sbp
oneflow.sbp.broadcast  # rank0 和 rank1 输出相同
```

doombeaker commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -0,0 +1,133 @@
+
+class TestToGlobalLocal(oneflow.unittest.TestCase):
+    placement = flow.placement('cpu', ranks=[0, 1])
+    sbp = (flow.sbp.broadcast,)
```
这里只做了 broadcast 的 SBP 测试，感觉覆盖还不够全。是否考虑把 S 合 Paritial 都加上？

Alive1024 replied:
已添加 Split 和 Partial 的测试

doombeaker commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -0,0 +1,133 @@
+
+    def test_any_input(test_case):
+        tensor = flow.zeros((3, 4))
+        tensor_lt = [flow.tensor([1, 2, 3]), flow.randn((2, 3, 4))]
+        tensor_tp = (flow.zeros((2, 2)), flow.ones((2, 3)), flow.randn((3, 5)))
+        tensor_dt = {'tensor': tensor, 'tensor_lt': tensor_lt}
```
```suggestion
        tensor_list = [flow.tensor([1, 2, 3]), flow.randn((2, 3, 4))]
        tensor_tuple = (flow.zeros((2, 2)), flow.ones((2, 3)), flow.randn((3, 5)))
        tensor_dict = {'tensor': tensor, 'tensor_lt': tensor_lt}
```

这里的用全称吧，用缩写但是又不是很常见，阅读的任反应不过来。其它相应位置都改一下。

Alive1024 replied:
已修改

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+def to_global_op(input, placement=None, sbp=None, **kwargs):
+    r"""Converts the input tensor or input tensor(s) in list/tuple/dict to global tensor(s).
+    
+    It performs the same conversion as :func:`oneflow.Tensor.to_global` to the tensor or every tensor in the input.
```
有道理，已删除这句

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+
+    For a tensor input: please refer to the examples in :func:`oneflow.Tensor.to_global`.
+
+    For an input of other type (take a state dict as an example):
```
这两个接口可以接受任何类型的输入，比如 int、str、None 等。不过只会对输入中包含的 Tensor (及其子类例如 Parameter) 进行操作。所以按 Tensor 和容器两种情况应该不太恰当？tensor 和 state dict 是最典型的两类输入，所以用这两个作为示例代码

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+
+def _to_local_tensor(input_tensor):
+    if not input_tensor.is_global:
+        warnings.warn("The tensor should be global, it will not be converted if not.")
```
已修正

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
 def to_local_op(input):
-    assert input.is_global, "input must be a global tensor!"
-    return flow._C.to_local(input)
+    r"""Returns the local component of the input.
```
是的，是这个意思。我先改成 part 了

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+
+    For a tensor input: please refer to the examples in :func:`oneflow.Tensor.to_local`.
+
+    For an input of other type (take a state dict as an example):
```
解释同上

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
-    return flow._C.to_local(input)
+    r"""Returns the local component of the input.
+
+    It performs the same conversion as :func:`oneflow.Tensor.to_local` to the tensor(s) in the input.
```
已删除这句

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+    It performs the same conversion as :func:`oneflow.Tensor.to_local` to the tensor(s) in the input.
+    
+    Returns:
+        The converted input.
```
确实称作 converted 不太恰当，此处暂时还没改。感觉这里对返回值的描述和开头对功能的描述其实是重合的？

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+    Returns:
+        The converted input.
+
+    For a tensor input: please refer to the examples in :func:`oneflow.Tensor.to_local`.
```
这里的引用是为了给出对 tensor 进行 to_local 操作时的示例代码，因为和调用 oneflow.Tensor.to_local 相比只是一个通过 oneflow.to_local(input)调用，一个在 Tensor 对象上调用，其余的都一样，所以没有重复再写一遍示例，而是链接到了那边。删除了的话就缺少关于 tensor 类型输入的示例了

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+            if isinstance(node._value, Tensor):
+                return _to_global_tensor(node._value, placement, sbp, **kwargs)
+            else:
+                warnings.warn("Non-Tensor type: {} encountered, it will not be converted.".format(type(node._value)))
```
已修正

Alive1024 commented on python/oneflow/nn/modules/global_cast.py:
```diff
@@ -74,17 +76,138 @@ def global_to_global_op(
+            if isinstance(node._value, Tensor):
+                return _to_local_tensor(node._value)
+            else:
+                warnings.warn("Non-Tensor type: {} encountered, it will not be converted.".format(type(node._value)))
```
已修正

Alive1024 commented on python/oneflow/test/modules/test_global_cast.py:
```diff
@@ -0,0 +1,133 @@
+
+
+class TestToGlobalLocal(oneflow.unittest.TestCase):
+    placement = flow.placement('cpu', ranks=[0, 1])
```
暂时没加，关于这个情况的处理还需要讨论确定

Alive1024 commented:
在进行 https://github.com/Oneflow-Inc/OneTeam/issues/1031#issuecomment-1109451227 所描述的测试 1 时遇到一个问题：
目前加载模型参数时的代码：

```python
if flow.env.get_rank() == 0:
   local_state_dict = flow.load(state_dict_dir)
else:
   local_state_dict = None

global_state_dict = flow.to_global(local_state_dict, placement=P, sbp=B)
```

执行到 `flow.to_global(local_state_dict, placement=P, sbp=B)` 时会卡住，显存仍有占用、但利用率为 0，没有任何报错，一直不会终止。

实验后发现，只要不同 rank 得到的 state dict 的值不同，就会在 to_global 的时候出现上述卡住的情况

Alive1024 commented:
debug 过程中发现 bug: nn.Graph 的 state dict 中的 `System-Train-TrainStep` 只被放在了在 rank 0 (https://github.com/Oneflow-Inc/OneTeam/issues/1371) 此问题待解决

Alive1024 commented:
1. 之前使用到的 `IONode` 被重构为了 `ArgsTree` ，目前已经相应地替换。但 `ArgsTree` 只允许  `dict` `tuple` `list`  三种类型的输入，因此不能再像之前那样只要不是 tensor 就统一构造一个 `IONode`，如果继续支持  `int` `str`  等其他类型的输入的话，还需要额外处理下 （本质上对于这些值就是 warning + 直接返回）。想确认下这样做是不是恰当的？

2. 目前 `oneflow.to_global` 的行为是：
当在 rank 0 上的输入为一个 OrderedDict、其他 rank 上的输入为 None 的时候，将 rank 0 的 state dict 进行广播，然后再进行原来 oneflow.to_global() 的过程。这样的话就能正常执行之前这里设想的[“测试1”](https://github.com/Oneflow-Inc/OneTeam/issues/1031#issuecomment-1109451227) 。是只把这种情况当作一种特殊的 case 在 `oneflow.to_global`  中进行判断和处理？目前判断这种特殊的输入情况的条件还不太恰当，或者说可以给 `oneflow.to_global` 增加一个参数？

strint commented:
> 如果继续支持 int str 等其他类型的输入的话，还需要额外处理下 （本质上对于这些值就是 warning + 直接返回）。想确认下这样做是不是恰当的？

是没问题的，之前只是提供了类型，ArgsTree里面需要自己看一下类型，然后根据类型做处理就好

strint commented:
> 其他 rank 上的输入为 None

可以先拿 rank 0 作为测试。

测试成功后再改的更为通用，就是拿非 None 的，向 None 的 Rank 进行发送，应该是更通用的。这里其实有个准确的判断标准，就是 placement 里面包括的 rank，向 placement 里面不包括的 rank，做发送。

strint commented on python/oneflow/test/graph/test_graph_save_load_global_new_version.py:
```diff
@@ -0,0 +1,166 @@
+
+
+# @unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+# @flow.unittest.skip_unless_1n2d()
```
这两个条件判断还是需要加上，一个判断有 cuda 设备时执行，一个判断2卡时执行

strint commented on python/oneflow/test/graph/test_graph_save_load_global_new_version.py:
```diff
@@ -0,0 +1,166 @@
+            else:
+                local_state_dict = None
+
+            global_state_dict = flow.to_global(local_state_dict, placement=P, sbp=B, src_rank=0)
```
这个 src_rank 应该可以消除掉，办法是用 placement，用一个 placement 内的合法 rank 向其它 rank 同步 dict

Alive1024 commented:
给 `flow.to_global` 增加了广播的逻辑后，对于 Global Tensor 类型的输入，当 SBP 的类型为 split 或 partial_sum 时，会卡住，例如：
```python
import oneflow as flow

P = flow.placement("cpu", ranks=[0, 1])
SBP = flow.sbp.split(0)
# SBP = flow.sbp.partial_sum

global_x = flow.ones((2, 2), placement=P, sbp=SBP)
flow.to_global(global_x, placement=P, sbp=SBP)    # 此处会卡住
```

strint commented:
```
P = flow.placement("cpu", ranks=[0, 1])
SBP = flow.sbp.split(0)
global_x = flow.ones((2, 2), placement=P, sbp=SBP)
flow.to_global(global_x, placement=P, sbp=SBP)
```

这里 to_global 的输入 global_x 的 sbp 和要转换的的 sbp 是一样的，那这里按说就是一个可以直接返回的操作。

在两个 rank 上分别打印下 global_x 和 global_x.to_local() 看看。

另外也试下这么写：
```
P = flow.placement("cpu", ranks=[0, 1])
SBP = flow.sbp.split(0)
global_x = flow.ones((2, 2), placement=P, sbp=SBP)
global_y = global_x.to_global(placement=P, sbp=SBP)
```
看这样是否有问题。

Alive1024 commented:
目前按 https://github.com/Oneflow-Inc/OneTeam/issues/1586 中实现的 `oneflow.dict_to_global` ：

```python
def dict_to_global(input_dict, placement, sbp, *, sbp_for_special_keys):
    assert isinstance(input_dict, dict) or input_dict is None, "The input must be a dict or None!"
    
    if input_dict is not None:
        for key, value in input_dict.items():
            if isinstance(value, dict):
                # Update the key(s) in sbp_for_special_keys (removing possible prefix)
                for special_key in sbp_for_special_keys.keys():
                    if special_key.startswith(key):
                        sbp_for_special_keys[special_key.replace(key + ".", "")] = sbp_for_special_keys.pop(special_key)
                dict_to_global(value, placement, sbp, sbp_for_special_keys=sbp_for_special_keys)
            elif isinstance(value, Tensor):
                if key not in sbp_for_special_keys.keys():
                    input_dict[key] = _to_global_tensor(input_dict[key], placement=placement, sbp=sbp)
                else:
                    input_dict[key] = _to_global_tensor(input_dict[key], placement=placement, sbp=sbp_for_special_keys[key])
            else:
                warnings.warn("Non-Tensor type: {} encountered, it will remain the same.".format(type(value)))

    return input_dict
```

即针对 state dict 中不同的 key 对应的 tensor 使用不同的 SBP 进行 to_global 的操作，


### 例子

假设目前在 4 个 ranks 上进行流水并行，`model_file_placement` 为 `oneflow.placement("cpu", ranks=[0, 1])`。

**保存模型：**

```python
iter1_state_dict = graph_model.state_dict()
model_file_state_dict = flow.dict_to_global(iter1_state_dict, placement=model_file_placement, sbp=flow.sbp.split(0),
                                            sbp_for_special_keys=sbp_for_special_keys)
if flow.env.get_rank() in model_file_placement.ranks:
    flow.save(flow.to_local(model_file_state_dict), state_dict_dir)
```

**加载模型：**

```python
if cur_rank in model_file_placement.ranks:
    local_state_dict = flow.load(state_dict_dir)
else:
    local_state_dict = None

global_state_dict = flow.dict_to_global(local_state_dict, placement=model_file_placement, sbp=flow.sbp.split(0),
                                        sbp_for_special_keys=sbp_for_special_keys)
graph_model.load_state_dict(global_state_dict)
```

### 错误

保存时是没有问题的，但在加载时执行到 ` flow.dict_to_global` 时就会卡住。错误出在广播的问题上。
按之前 `oneflow.to_global` 中的想法，将 `model_file_placement.ranks` 中的第一个 rank 作为 `src_rank`，广播 `src_rank` 上的 data structure 到其他 rank(s)。


### 尝试的解决办法

将 `oneflow.dict_to_global` 中的 `_to_global_tensor` 换为带有广播机制的 `oneflow.to_global` 时，也是会卡住，因为广播时使用的 C++ 层的 `oneflow._oneflow_internal.cpu_broadcast` 必须在所有 ranks 上执行，否则会处于等待状态；进而尝试在每次对 tensor 进行 to_gloabal 之前进行广播，但操作 tensor 是在 `model_file_placement.ranks`上进行的，需要其他 ranks 做出响应，这似乎很难实现。

strint commented:
Finished in: https://github.com/Oneflow-Inc/oneflow/pull/8868
